controllers/kubeadmconfig_controller.go

```diff
@@ -130,10 +133,13 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 			return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
 		}
 
-		//TODO(fp) use init lock so only the first machine configured as control plane get processed, everything else gets requeued
-
-		// otherwise it is a init control plane
-		// Nb. in this case JoinConfiguration should not be defined by users, but in case of misconfigurations, CAPBK simply ignore it
+		// acquire the init lock so that only the first machine configured
+		// as control plane get processed here
+		// if not the first, requeue
+		if !initLocker.Acquire(cluster) {
+			log.Info("A control plane is already being initialized, requeing until control plane is ready")
+			return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
+		}
 
```

Comment:
You might want to `defer initLocker.Release(cluster)` here and remove the other calls, it should simplify a lot the code and avoid issues down the line :)

Reply:
it looks like not *every* exit case needs to release the lock, just the errors?

Reply:
> it looks like not _every_ exit case needs to release the lock, just the errors?

You are correct Chuck. Only the error cases need to release the lock.

Reply:
this could be:
```
releaseLock := true
...

if !initLocker.Acquire(cluster) {
	log.Info("A control plane is already being initialized, requeing until control plane is ready")
	releaseLock = false
	return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
}
...

defer func() {
	if releaseLock {
		someConditioninitLocker.Release(cluster)
	}
)
```

Reply:
@neolit123 the problem is still the same. Either we have to remember to release the lock at all the error return sites or remember to set releaseLock=true at all the non-error return sites.

Reply:
my understanding is that the only branch that needs to set `releaseLock = false` is `if !initLocker.Acquire(cluster) {`

Reply:
This function is using named returns. Are you able to detect that there's an error and have defer release the lock based off that? 

```
defer func() {
  if rerr != nil {
    initLocker.Release(cluster)
  }
}
```

Reply:
Another option could be:
Pull that Locker logic into another function. There's a beginning and end to that locker logic anyways.

It could be something like this: 
```
func LockerLogic(...) err { // or whatever you want to name it
}

func Reconcile(...) {
  
  initLocker.Acquire(cluster)
  err := LockerLogic()
  initLocker.Release(cluster)
  if err != nil {
    return ctrl.Result{}, err
  }
```

}